### ISSUE TITLE
containerd: ensure we get `container.SetGraceTime` right

### ISSUE THREAD

**Issue by cirocosta** [concourse/concourse#5134]
### What challenge are you facing? 

`container.SetGraceTime` and `container.GraceTime` **should not** be no-ops: currently, deletion of hijacked containers suffer as a result.

From the set of methods that were implemeted iin #5114 with regards to container
lifecycle, one is currently missing: `SetGraceTime`.

It turns out that we use this method when marking for destruction a container
that has been hijacked into:

https://github.com/concourse/concourse/blob/54eb47b4e2ac164435f8d8dc4a68c2c8feb6998c/atc/gc/container_collector.go#L18
https://github.com/concourse/concourse/blob/54eb47b4e2ac164435f8d8dc4a68c2c8feb6998c/atc/gc/container_collector.go#L233-L237

looking at garden server implementation (which wraps our implentation of a
Backend as an actual HTTP server), it *seems* like it takes care of dealing with
the grace time by itself and then scheduling `Destroy`s to be called:

```go
func (s *GardenServer) handleSetGraceTime(w http.ResponseWriter, r *http.Request) {
	handle := r.FormValue(":handle")

	var graceTime time.Duration

	// ...

	container, err := s.backend.Lookup(handle)
	if err != nil {
		s.writeError(w, err, hLog)
		return
	}


	container.SetGraceTime(graceTime)
	s.bomberman.Reset(container)
	s.writeSuccess(w)
}
```

https://github.com/cloudfoundry/garden/blob/31a0ffe798632d210f8c5f64379ed4464e1fa410/server/request_handling.go#L698-L721

which means that it *might* be that there's not really much on our side left to
be done aside from keeping track of that value (and thus returning it
accordingly when asked for).

```go
func (s *GardenServer) handleCreate(w http.ResponseWriter, r *http.Request) {
	var spec garden.ContainerSpec
	if !s.readRequest(&spec, w, r) {
		return
	}

	container, err := s.backend.Create(spec)
	if err != nil {
		s.writeError(w, err, hLog)
		return
	}

	//
	s.bomberman.Strap(container)
}

func (b *Bomberman) Strap(container garden.Container) {
	b.bomb <- bomb{Action: strap, Container: container}
}

func (b *Bomberman) manageBombs () {
	for {
		select {
		case bombSignal := <-b.bomb:
			switch bombSignal.Action {
			case strap:
				container := bombSignal.Container

				bomb := timebomb.New(

					// [cc] retrieves grace time here
					// 
					b.backend.GraceTime(container),

					func() {
						b.detonate(container)
						b.cleanup <- container.Handle()
					},
				)


				timeBombs[container.Handle()] = bomb
				if b.backend.GraceTime(container) != 0 {
					bomb.Strap()
				}

		// ...
}
```

https://github.com/cloudfoundry/garden/blob/31a0ffe798632d210f8c5f64379ed4464e1fa410/server/request_handling.go#L61-L93
https://github.com/cloudfoundry/garden/blob/31a0ffe798632d210f8c5f64379ed4464e1fa410/server/bomberman/bomberman.go#L49-L51
https://github.com/cloudfoundry/garden/blob/31a0ffe798632d210f8c5f64379ed4464e1fa410/server/bomberman/bomberman.go#L80-L81

### What would make this better?

Have those two methods properly implemented.

**Comment by cirocosta** [comment]
@muntac @aoldershaw it might be worth taking a look at how https://github.com/vito/houdini and https://github.com/cloudfoundry/guardian implemented `container.SetGraceTime`

**Comment by aoldershaw** [comment]
Without implementing `GraceTime` and `SetGraceTime`, the Garden server will always see the grace time as 0.

https://github.com/cloudfoundry/garden/blob/31a0ffe798632d210f8c5f64379ed4464e1fa410/server/bomberman/bomberman.go#L109-L110

The "timebomb" is then reset with a countdown of 0 if it had a timer from earlier (`b.timer != nil`), or if the timer is not paused (`b.pauses == 0`):

https://github.com/cloudfoundry/garden/blob/31a0ffe798632d210f8c5f64379ed4464e1fa410/server/timebomb/timebomb.go#L86-L88

`b.timer` will always be `nil` (since the grace time is 0 when the container is created, and it is only created if it is non-zero), and so the timebomb will be reset if and only if it is not paused.

If the timebomb is not paused, it will immediately "detonate" (we run `time.AfterFunc(b.countdown, b.detonate)` where `b.countdown` is `0` since we did not properly set the grace time).

https://github.com/cloudfoundry/garden/blob/31a0ffe798632d210f8c5f64379ed4464e1fa410/server/bomberman/bomberman.go#L88-L90

If the timebomb is paused, it will never detonate. We don't reset the timer during the `SetGraceTime` call since `b.pauses == 0`. Normally, the timebomb will be reset when it is unpaused. However, since `b.countdown == 0`, we never hit that case here:

https://github.com/cloudfoundry/garden/blob/31a0ffe798632d210f8c5f64379ed4464e1fa410/server/timebomb/timebomb.go#L71-L73

The timebomb is paused when we hijack into the container, and unpaused when we exit that shell.

---

TLDR; if we don't implement `container.SetGraceTime` and `backend.GraceTime` properly, containers that are currently being hijacked at the time of a GC sweep will never be cleaned up.

Additionally, we'll want a 5 minute grace time so that hijacked containers aren't removed immediately after exiting the hijack shell (in case a user accidentally exited the shell but wants to re-hijack the container).
